### PR TITLE
2048

### PR DESCRIPTION
**Pull request by matsarat**

### CODE REVIEW

File: 2048.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        move_down(board)
+        sum_down(board)
+        move_down(board)
+
+
+def game(number_of_rows, number_of_columns):
+    board = create_board(number_of_rows, number_of_columns)
+    instructions()
+    place_two_initial_elements_on_board(board)
+    print_board(board)
```
Comment (Czeffik) [inline 2048.py]: you don't need this method here

File: 2048.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    while game_over(board) == False:
+        valid_players_move(board)
+        get_fields_with_zero_value(board)
```
Comment (Czeffik) [inline 2048.py]: you don't need this method here

File: 2048.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        game_over(board)
```
Comment (Czeffik) [inline 2048.py]: you don't need this method here

File: 2048.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+def game_over(board):
+    winning_number = 2048
+    for row in board:
+        if winning_number in row:
+            print("You did it!")
+            return True
+    if len(get_fields_with_zero_value(board)) > 0:
+        return False
+    if len(get_fields_with_zero_value(board)) == 0:
```
Comment (Czeffik) [inline 2048.py]: you don't need this if here

File: 2048.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    next_field.value = 2
+
+
+def game_over(board):
+    winning_number = 2048
+    for row in board:
+        if winning_number in row:
+            print("You did it!")
+            return True
+    if len(get_fields_with_zero_value(board)) > 0:
```
Comment (Czeffik) [inline 2048.py:192]: you should add here `or` when we have on board situation where two fields have same value

**Review by Czeffik — CHANGES_REQUESTED** [review]

File: 2048.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+
+def instructions():
+    print("To move left - press 'a'")
+    print("To move right - press 'd'")
+    print("To move down - press 's'")
+    print("To move up - press 'w'")
+
+
+def place_two_initial_elements_on_board(board):
```
Comment (Czeffik) [inline 2048.py:52]: use `def get_fields_with_zero_value(board)` here

File: 2048.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+def get_fields_with_zero_value(board):
+    fields_with_zero_value = []
+    for row in board:
+        for field in row:
+            if field.value == 0:
+                fields_with_zero_value.append(field)
+    return fields_with_zero_value
+
+def place_next_element_on_board(board):
+    next_field = random.choice(get_fields_with_zero_value(board))
```
Comment (Czeffik) [inline 2048.py]: `random.choice()` docs:
```doc
Return a random element from the non-empty sequence seq. If seq is empty, raises IndexError.
```
IMHO with current logic this method can raise `IndexError`

File: 2048.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+        print("YOU LOST!")
+        return True
+
+
+def valid_players_move(board):
+    players_move = str(input("Next move: ").lower())
+    if players_move == "a":
+        move_left(board)
+        sum_left(board)
+        move_left(board)
```
Comment (Czeffik) [inline 2048.py:223]: hmm... If I remember our design of this method it didn't look like this implementation :D Check your notes!

**Review by Czeffik — CHANGES_REQUESTED** [review]

File: 2048.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+def place_two_initial_elements_on_board(board):
+    initial_element_value = 2
+    get_fields_with_zero_value(board)
+
+    first_initial_element = random.choice(get_fields_with_zero_value(board))
+    first_initial_element.value = initial_element_value
+
+    get_fields_with_zero_value(board)
+    second_initial_element = random.choice(get_fields_with_zero_value(board))
+    second_initial_element.value = initial_element_value
```
Comment (Czeffik) [inline 2048.py]: use `def get_fields_with_zero_value(board)` here

File: 2048.py [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+def check_if_won(board):
+    winning_number = 16
+    for row in board:
+        for field in row:
+            if field.value == winning_number:
+                return True
+    return False
+
+
+def get_temporary_board_values(board):
```
Comment (Czeffik) [inline 2048.py]: What is the difference between these two methods: `get_temporary_board_values` and `get_board_values`?

File: 2048.py [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+
+def validate_players_move(board):
+    temporary_values_board = get_temporary_board_values(board)
+    players_move = str(input("Next move: ").lower())
+    if players_move == "a":
+        move_left(board)
+        sum_left(board)
+        move_left(board)
+        values_board = get_board_values(board)
+        if temporary_values_board == values_board:
```
Comment (Czeffik) [inline 2048.py]: this sequence is repeated 4 times!!! What can you do with it? :D

File: 2048.py [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+        else:
+            place_next_element_on_board(board)
+
+
+def game(number_of_rows, number_of_columns):
+    board = create_board(number_of_rows, number_of_columns)
+    instructions()
+    place_two_initial_elements_on_board(board)
+    print_board(board)
+    while game_over(board) == False or check_if_similar_neighboring(board) == True:
```
Comment (Czeffik) [inline 2048.py]: invoke `check_if_similar_neighboring(board)` inside `gamve_over` method. Also you can have: `while not game_over(board)` which is easier to read

File: 2048.py [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+
+
+
+def game_over(board):
+    if len(get_fields_with_zero_value(board)) > 0:
+        return False
+
+
+def check_if_won(board):
+    winning_number = 16
```
Comment (Czeffik) [inline 2048.py]: hmm :D 16? :D

File: 2048.py [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+    print_board(board)
+    while game_over(board) == False or check_if_similar_neighboring(board) == True:
+        validate_players_move(board)
+        if check_if_won(board) == True:
+            print("You did it!")
+            print_board(board)
+            break
+        print_board(board)
+
+game(3, 3)
```
Comment (Czeffik) [inline 2048.py]: 4,4

File: 2048.py [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+            print("INVALID MOVE!")
+        else:
+            place_next_element_on_board(board)
+
+
+def game(number_of_rows, number_of_columns):
+    board = create_board(number_of_rows, number_of_columns)
+    instructions()
+    place_two_initial_elements_on_board(board)
+    print_board(board)
```
Comment (Czeffik) [inline 2048.py]: with current logic this print for sure is not necessary! You can `print_board` firstly in `while` loop

**Review by Czeffik — CHANGES_REQUESTED** [review]

File: 2048.py (hunk 10 above)
Comment (matsarat) [reply 2048.py]: One of these got deleted and correction in method previously using both of them was made.

File: 2048.py (hunk 11 above)
Comment (matsarat) [reply 2048.py]: Method was extracted and called check_if_move_is_valid .

File: 2048.py (hunk 12 above)
Comment (matsarat) [reply 2048.py]: Done.

File: 2048.py (hunk 13 above)
Comment (matsarat) [reply 2048.py]: Convenient during tests. Corrected.

File: 2048.py (hunk 14 above)
Comment (matsarat) [reply 2048.py]: Convenient during tests. Corrected.

File: 2048.py (hunk 15 above)
Comment (matsarat) [reply 2048.py]: Corrected. Also added printing board after losing game in game_over method.

File: 2048.py [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+
+
+def game(number_of_rows, number_of_columns):
+    board = create_board(number_of_rows, number_of_columns)
+    instructions()
+    place_two_initial_elements_on_board(board)
+    while not game_over(board):
+        print_board(board)
+        validated_players_move(board)
+        if check_if_won(board) == True:
```
Comment (Czeffik) [inline 2048.py:260]: FYI - this method also can be moved to `game_over` method (minor - you do not have to do that)

File: 2048.py [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+        place_next_element_on_board(board)
+
+
+def game(number_of_rows, number_of_columns):
+    board = create_board(number_of_rows, number_of_columns)
+    instructions()
+    place_two_initial_elements_on_board(board)
+    while not game_over(board):
+        print_board(board)
+        validated_players_move(board)
```
Comment (Czeffik) [inline 2048.py:259]: FYI - for me it would be clearer when this method would be named `make_move` (minor - you do not have to do that)

**Review by Czeffik — APPROVED** [review]

**Pull request merged**